File: 19-remove-nth-node-from-end-of-list/remove-nth-node-from-end-of-list.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

function removeNthFromEnd(head: ListNode | null, n: number): ListNode | null {
    if(n === 0) return head;
    if(head === null || head.next === null) return null;
    let count = 0;
    let curHead = head;
    while(curHead != null){
        count ++;
        curHead = curHead.next;
    }
    let requiredNodePos = count - n;
    if(requiredNodePos <= 0) return head.next;
    curHead = head;
    for(let i = 1; i < requiredNodePos; i ++){
        curHead = curHead.next;
    }
    curHead.next = curHead.next.next;
    return head;
};

File: 19-remove-nth-node-from-end-of-list/remove-nth-node-from-end-of-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
      if(n < 0) return head;
      int count = 0;
      ListNode curNode = head;
      while(curNode != null){
        count ++;
        curNode = curNode.next;
      }
      if(n > count) return head;
      if(n == count) return head.next;
      curNode = head;
      int num = count - n - 1;
      while(num > 0){
        num --;
        curNode = curNode.next;
      }
      curNode.next = curNode.next.next;
      return head;
    }
}